src/providers/ipa/ipa_subdomains_ext_groups.c

```diff
@@ -929,25 +943,26 @@ static void ipa_add_trusted_memberships_get_next(struct tevent_req *req)
     struct ldb_dn *group_dn;
     int ret;
     const struct ldb_val *val;
-    bool missing_groups;
     const char *fq_name;
     char *tmp_str;
 
-    while (state->groups[state->iter] != NULL
-            && state->groups[state->iter][0] == '\0') {
+    while (state->missing_groups[state->iter] != NULL
+            && state->missing_groups[state->iter][0] == '\0') {
         state->iter++;
     }
 
-    if (state->groups[state->iter] == NULL) {
-        ret = add_ad_user_to_cached_groups(state->user_dn, state->user_dom,
+    if (state->missing_groups[state->iter] == NULL) {
+        talloc_zfree(state->missing_groups);
+        ret = add_ad_user_to_cached_groups(state, state->user_dn,
```

Comment:
> Since add_ad_user_to_cached_groups() is called twice

Why?

Reply:
Hi,

even before the recent changes the first run would return a list of group missing in the cache which have to be read from LDAP first before the membership for those can be added in the second run. If the cache is already used for some time, chances are that all relevant groups are already cached and only a single run is needed.

bye,
Sumit